sysl2/sysl/validate.go

```diff
@@ -0,0 +1,377 @@
+	case *sysl.Type_TypeRef:
+		if t.TypeRef.GetRef().GetAppname() != nil {
+			return t.TypeRef.GetRef().GetAppname().GetPart()[0]
+		}
+		return t.TypeRef.GetRef().GetPath()[0]
+	default:
+		return "Unknown"
+	}
+}
+
+func isCollectionType(syslType *sysl.Type) bool {
+	switch syslType.Type.(type) {
+	case *sysl.Type_Set, *sysl.Type_Sequence, *sysl.Type_List_, *sysl.Type_Map_:
+		return true
+	default:
+		return false
+	}
+}
+
+func validateEntryPoint(views map[string]*sysl.View, start string) []ValidationMsg {
+	view, exists := views[start]
+
+	if !exists {
+		return []ValidationMsg{{
+			Message: fmt.Sprintf("Entry point view: '%s' is undefined", start),
+			MsgType: ERROR,
+		}}
+	}
+
+	if getTypeName(view.GetRetType()) != start || isCollectionType(view.GetRetType()) {
+		return []ValidationMsg{{
+			Message: fmt.Sprintf("Output type of entry point view: '%s' should be '%s'", start, start),
+			MsgType: ERROR,
+		}}
+	}
+
+	return nil
+}
+
+func validateFileName(views map[string]*sysl.View) []ValidationMsg {
+	viewName := "filename"
+	view, exists := views[viewName]
+	validationMsgs := make([]ValidationMsg, 0, 2)
+
+	if !exists {
+		return []ValidationMsg{{
+			Message: "View 'filename' is undefined",
+			MsgType: ERROR,
+		}}
+	}
+
+	if getTypeName(view.GetRetType()) != "STRING" || isCollectionType(view.GetRetType()) {
+		validationMsgs = append(validationMsgs, ValidationMsg{
+			Message: "In view 'filename', output type should be 'string'",
+			MsgType: ERROR,
+		})
+	}
+
+	assignCount := 0
+	for _, stmt := range view.GetExpr().GetTransform().GetStmt() {
+		if stmt.GetAssign() != nil {
+			if assignCount == 0 && stmt.GetAssign().GetName() != viewName {
+				validationMsgs = append(validationMsgs, ValidationMsg{
+					Message: "In view 'filename' : missing type: 'filename'",
+					MsgType: ERROR,
+				})
+			} else if assignCount > 0 {
+				validationMsgs = append(validationMsgs, ValidationMsg{
+					Message: fmt.Sprintf("In view 'filename' : Excess assignments: '%s'", stmt.GetAssign().GetName()),
+					MsgType: ERROR,
+				})
+			}
+			assignCount++
+		}
+	}
+	return validationMsgs
+}
+
+func getImplAttrNames(attrs map[string]*sysl.Type) map[string]struct{} {
+	implAttrNames := map[string]struct{}{}
+
+	for attrName := range attrs {
+		implAttrNames[attrName] = struct{}{}
+	}
+
+	return implAttrNames
+}
+
+func compareSpecVsImpl(grammarSpec, specAttrs, implAttrs map[string]*sysl.Type,
+	implAttrNames map[string]struct{}, viewName string) []ValidationMsg {
+	validationMsgs := make([]ValidationMsg, 0, len(implAttrs))
+
+	for gk, gv := range specAttrs {
+		if grammarSpec[gk].GetOneOf() != nil {
+			validationMsgs = append(validationMsgs, compareOneOf(grammarSpec, implAttrs, implAttrNames, viewName, gk)...)
+		} else if _, exists := implAttrs[gk]; !exists {
+			if !gv.GetOpt() {
+				validationMsgs = append(validationMsgs, ValidationMsg{
+					Message: fmt.Sprintf("In view '%s', type '%s' is missing", viewName, gk),
```

Comment:
Can I have this message:
`Missing required field (%s=gk) in  View (%s=viewName) with return Type (%s=typeName)`

Reply:
Return type is not present in this scope. Will require an additional function arg.